src/test/fuzz/process_messages.cpp

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <consensus/consensus.h>
+#include <net.h>
+#include <net_processing.h>
+#include <protocol.h>
+#include <test/fuzz/FuzzedDataProvider.h>
+#include <test/fuzz/fuzz.h>
+#include <test/fuzz/util.h>
+#include <test/util/mining.h>
+#include <test/util/net.h>
+#include <test/util/setup_common.h>
+#include <util/memory.h>
+#include <validation.h>
+#include <validationinterface.h>
+
+const RegTestingSetup* g_setup;
+
+void initialize()
+{
+    static RegTestingSetup setup{};
+    g_setup = &setup;
+
+    for (int i = 0; i < 2 * COINBASE_MATURITY; i++) {
+        MineBlock(g_setup->m_node, CScript() << OP_TRUE);
+    }
+    SyncWithValidationInterfaceQueue();
+}
+
+void test_one_input(const std::vector<uint8_t>& buffer)
+{
+    FuzzedDataProvider fuzzed_data_provider(buffer.data(), buffer.size());
+
+    ConnmanTestMsg& connman = *(ConnmanTestMsg*)g_setup->m_node.connman.get();
+    std::vector<CNode*> peers;
+
+    const auto num_peers_to_add = fuzzed_data_provider.ConsumeIntegralInRange(1, 3);
+    for (int i = 0; i < num_peers_to_add; ++i) {
+        const ServiceFlags service_flags = ServiceFlags(fuzzed_data_provider.ConsumeIntegral<uint64_t>());
+        const bool inbound{fuzzed_data_provider.ConsumeBool()};
+        const bool block_relay_only{fuzzed_data_provider.ConsumeBool()};
+        peers.push_back(MakeUnique<CNode>(i, service_flags, 0, INVALID_SOCKET, CAddress{CService{in_addr{0x0100007f}, 7777}, NODE_NETWORK}, 0, 0, CAddress{}, std::string{}, inbound, block_relay_only).release());
+        CNode& p2p_node = *peers.back();
+
+        p2p_node.fSuccessfullyConnected = true;
+        p2p_node.fPauseSend = false;
+        p2p_node.nVersion = PROTOCOL_VERSION;
+        p2p_node.SetSendVersion(PROTOCOL_VERSION);
+        g_setup->m_node.peer_logic->InitializeNode(&p2p_node);
+
+        connman.AddTestNode(p2p_node);
+    }
+
+    while (fuzzed_data_provider.ConsumeBool()) {
+        const std::string random_message_type{fuzzed_data_provider.ConsumeBytesAsString(CMessageHeader::COMMAND_SIZE).c_str()};
```

Comment:
Could the fuzzer check "interesting" cases more quickly by choosing a random but valid command, instead of generating a random string of length `COMMAND_SIZE`?

Reply:
Good question. I think there is a trade-off here between giving the fuzz engine as much freedom to explore as possible at a potentially slight cost of performance (citation needed, giving more freedom could also increase efficiency).

Your suggestion would involve enumerating all netmessage types in an array and then having the fuzz engine pick one from it. This is possible, but as soon as you add a new message type, the developed seeds (inputs) render invalid and need to be adjusted. Readjusting the seeds should not be an issue for the fuzz engine, but it has to be done.

Apart from the above slight inconvenience of having to readjust the seeds on every added (or removed) message type, there is also an issue that the enumeration could accidentally miss one message type. On top, Bitcoin Core ignores unknown message types, and I'd say we want to test this code path as well. (Have at least one unknown message type generated by the fuzzer). This can also be fixed by adding an unknown message type to the fixed array the fuzz engine can pick from. E.g. "foobarfoo".

Moreover, we might also want to test for slight modifications of the same message type and see if they affect behavior in any way. For example b"filteradd\x01add" vs b"filteradd\x02". Those two should never be treated as a valid command and obviously shouldn't crash Bitcoin Core. There should be unit and maybe other fuzz tests in place to check this, but I think allowing the fuzz engine in this test more freedom can't hurt.

Finally, while I don't know the internals of fuzz engines, I believe they are really smart and quick about figuring out "interesting" strings that appear in the source code. If I had to guess what the performance penalty was of allowing any message type as opposed to limiting it to a fixed set of types, I'd presume it is an unnoticeable difference, maybe slight slow down or possibly a slight speed up even. A benchmark might help here.

I tried a few other presumed performance improvements while writing the test and surprisingly I found that writing the least code, simplest code, while allowing the fuzz engine the most freedom was also the fastest way to find the CVE.

Reply:
Thanks for the really thorough explanation!  That tradeoff makes sense to me, especially since I didn't know that fuzz engines might do better than "just generate random strings".

Reply:
I agree as well that the merged version makes sense, but just out of interest I ran both versions (PR and a version which samples an array with all `NetMsgType`s currently defined in protocol.cpp ) in parallel with libfuzzer and it was the altered version which found the reintroduced CVE first after 2.3M iterations / 2 hours, while the PR version hadn't found it after 10M tries. 
I found it interesting that while the `cov` metric was comparable between the runs, the `ft` metric of the fuzzer was consistently higher for the array version throughout the run.

Although this was just sample size 1, so finding the CVE might have just been a lucky mutation in the fuzzer algorithm...

Reply:
How many workers did you run in parallel?

Reply:
Just 1 for each version.

Reply:
This sounds like a promising speed-up. I wouldn't object either having this one replaced or a new fuzz target with the array version.

Reply:
I'll do some more runs tomorrow, maybe I was just lucky. I don't have a good feeling for the variance of runtimes yet.

Reply:
I saw it vary by a factor or 5 or 10 in my tests last week, so it could very well be just noise